### ISSUE TITLE
If tide is enabled on a repo, the tide status context should be required by branch protection

### ISSUE THREAD

**Issue by cblecker** [kubernetes/test-infra#6289]
/area prow
cc: @fejta @spxtr

**Comment by cblecker** [comment]
@fejta Any update on this?

**Comment by fejta** [comment]
/unassign
/assign @cjwagner 
/area prow

Making the tide context required breaks tide.

**Comment by cblecker** [comment]
Wouldn't this be a race condition, then?

What about for repos where there isn't other prow testing? If it's not required, then anyone with write access would be able to manually merge the PR as long as the CLA context passes (this is why we require the SQ context itself)

**Comment by fejta** [comment]
True true, which are both good reasons for considering to address it rather than closing... although removing everyone's write access would solve that problem :)

Cole has the most state here and will let him weigh in.

**Comment by cjwagner** [comment]
Making the tide context required doesn't work well with Tide, but it doesn't completely break us. It just causes us to (almost always) fail the first attempt at every merge and have to retry the next sync loop.

The issue is that statuses are updated from a separate controller thread instead of being updated as part of the main sync controller loop and that statuses are updated with lower priority than syncing (since we care more about actually merging things than updating statuses correctly).
In order to allow the tide status to be required, the sync controller would need to wait for the status to be updated to green before merging the PR, otherwise the merge fails because a required status is not green.

An easy, but mildly hacky solution would be to have the sync controller set the status to green before merging when necessary. There is still potential for a race with the status controller (especially if the status controller is 'lapped' by the sync controller), but i think a low chance of setting a red status on a merged PR is a better alternative to failing the first attempt at every merge.

**Comment by spiffxp** [comment]
/area prow/tide
Since we're paying attention to status contexts, is this worth address or do we close this as good enough for now?

**Comment by cblecker** [comment]
I would *still* really like this fwiw

**Comment by stevekuznetsov** [comment]
Isn't this as noted by Cole fundamentally going to introduce weird races with the two `tide` controllers and therefore very unlikely to actually happen?

**Comment by cblecker** [comment]
/remove-lifecycle rotten

The biggest use case for this is where you want to require using tide for merging, and not show the green button when tests are passing, but other conditions (like required/missing labels) aren't met.

**Comment by sftim** [comment]
I'd like to be able to single out PRs that were merged using administrative overrides or similar. To help with that filtering, it would be really handy if Tide marked PRs as passing as soon as they are ready to merge. At that point the same Tide  _or another one_ could merge the PR. So long as Git / GitHub don't allow the same PR to merge twice, I don't see how there's a data race problem there.

**Comment by sftim** [comment]
_Also_ as someone who now has the ability to override Prow / Tide, I'd like this privilege to be protected by the usual GitHub safety features. With the website repo, there are usually no tests failing so I'm fewer clicks away from merging something PR manually when I didn't mean to.

**Comment by sftim** [comment]
/remove-lifecycle rotten

For the final merge once Tide is ready to doit, I wonder if Tide could implement merging by:
- turning on [automerge](https://twitter.com/github/status/1336360682221133827) for that PR
   - I'm assuming you can enable it per PR
- marking the Tide merge blocking status as passed

**Comment by alvaroaleman** [comment]
This here is blocked by https://github.com/kubernetes/test-infra/issues/16472

The tide status context is only eventual consistent, so relying just on it is unsafe. Furthermore, there are situations when merging was technically ok because a PR does have passing tests against the lastest base branch revision, but would slow down overall merge rate as it would invalidate the results of a currently running batch test.

As an aside, I don't see much value in replacing the one api call to actually merge a PR in Tide with a Github feature while still continuing to use Tide.

**Comment by fejta** [comment]
I am in favor of closing this PR as working as intended.

**Comment by alvaroaleman** [comment]
#16472 got resolved, the next step here would be to:
* Write code that can deduce from the tide config if tide is enabled for a given repo
* If yes, makes the branch protector protect the tide status (initially protected via a feature gate)
